tempo_embeddings/embeddings/weaviate_database.py

```diff
@@ -0,0 +1,278 @@
+import json
+import logging
+import os
+import platform
+from typing import Any
+from typing import Iterable
+from typing import TypeVar
+import weaviate
+import weaviate.classes as wvc
+from transformers import AutoTokenizer
+from weaviate.classes.query import Filter
+from weaviate.classes.query import MetadataQuery
+from weaviate.util import generate_uuid5
+from ..text.corpus import Corpus
+from ..text.highlighting import Highlighting
+from ..text.passage import Passage
+from .vector_database import VectorDatabaseManagerWrapper
+
+
+Collection = TypeVar("Collection")
+logger = logging.getLogger(__name__)
+
+class WeaviateDatabaseManager(VectorDatabaseManagerWrapper):
+    """A Weaviate Database can create N collections and will always use THE SAME embedder function and tokenizer for all collections.
+    To create collections with different embedders one needs to create separate databases
+    """
+
+    def __init__(
+        self,
+        db_path: str = "weaviate_default_db",
+        embedder_name: str = None,
+        embedder_config: dict[str, Any] = None,
+        batch_size: int = 8,
+    ):
+        super().__init__(batch_size)
+        self.db_path = db_path
+        self.embedder_name = embedder_name
+        self.embedder_config = embedder_config or {"type": "default"}
+        self.tokenizer = AutoTokenizer.from_pretrained(embedder_name) if embedder_name else None
+        self.model = None
+        self.client = None
+        self.weaviate_headers = {}
+        
+        if self.embedder_config.get("type") == "hf":
+            self.embedding_function = wvc.config.Configure.Vectorizer.text2vec_huggingface(model=self.embedder_name)
+            self.weaviate_headers = {"X-HuggingFace-Api-Key": self.embedder_config["api_key"]}
+        elif self.embedder_config.get("type") == "custom_model" or self.embedder_config.get("type") == "default":
+            try:
+                self.model = self.embedder_config["model"]
+                self.model.batch_size = self.batch_size
+                self.embedding_function = wvc.config.Configure.Vectorizer.none()
+            except KeyError:
+                logger.error("If the type is 'custom_model' or 'default' you should pass the model object under Key 'model'")
+        else:
+            raise ValueError(f"Malformed embedder_config {self.embedder_config}. Check that 'type', 'api_key' and 'model' keys are properly populated.")
+
+        if os.path.exists(self.db_path):
+            self._load_config()
+        else:
+            os.makedirs(self.db_path)
+            self.config = {
+                "embedder_name": self.embedder_name,
+                "embedder_type": self.embedder_config["type"],
+                "existing_collections": [],
+            }
+        
+        
+    def connect(self):
+        with weaviate.connect_to_local() as client:
```

Comment:
For my understanding: this looks like it is supposed to connect to a local database instance. If that is correct, does it mean we need to add functionality for connecting to a remote database once we should need that?
Perhaps add a TODO comment for clarification.
```suggestion
        with weaviate.connect_to_local() as client:
            # TODO: add functionality for connecting to remote database
```

Reply:
Yes the comment is a good idea! Also this method is "fake" to keep consistency with the chroma functionality. For weaviate the connection is opened and closed everytime the API is called. In this case connect() only acts as a ping to know if the database is running